### PR TITLE
Flash: Add fees to update TXs before publishing them 

### PR DESCRIPTION
**Pull request by omerfirmak**
Part of #2200

### CODE REVIEW

File: source/agora/script/Signature.d [hunk 1]
```diff
@@ -142,39 +213,147 @@ unittest
...
+    import std.exception : assumeWontThrow;
+
+    if (sig_hash != SigHash.All)
+    {
+        assert(input_idx < tx.inputs.length, "Input index is out of range");
+        if (sig_hash & SigHash.Single)
+            assert(output_idx < tx.outputs.length, "Output index is out of range");
+    }
+
+    Transaction dup = () @trusted { return *cast(Transaction*)&tx; }();
```
Comment (Geod24) [inline source/agora/script/Signature.d:239]: We really need to get rid of this...

File: source/agora/flash/UpdateSigner.d [hunk 2]
```diff
@@ -480,17 +483,19 @@ public class UpdateSigner
         // Note that an update tx with seq 0 do not exist.
         if (this.seq_id == 0)
         {
-            return sign(this.kp.secret, this.conf.pair_pk, nonce_pair_pk,
-                priv_nonce.update.v, update_tx.getChallenge());
+            const sig_hash = cast(SigHash)(SigHash.Single | SigHash.AnyoneCanPay);
```
Comment (Geod24) [inline source/agora/flash/UpdateSigner.d]: Nope, a flag is not an enum. I wish D had something like C#'s flag, but it currently doesn't.
Casting an out-of-range value to an `enum` type is like taking a chainsaw to the type system, and I'd rather we do not do that.

File: source/agora/flash/UpdateSigner.d [hunk 3]
```diff
@@ -687,15 +698,15 @@ public class UpdateSigner
...
     {
         // if the current sequence is 0 then the update tx is a trigger tx that
         // only needs a multi-sig and does not require a sequence.
         // an update tx with seq 0 do not exist.
-        if (this.seq_id == 0)
+        if (seq_id == 0)
             return genKeyUnlock(update_multi_sig);
         else
-            return createUnlockUpdate(update_multi_sig, this.seq_id);
+            return createUnlockUpdate(update_multi_sig, seq_id);
```
Comment (Geod24) [inline source/agora/flash/UpdateSigner.d]: Minor, but that `else` is superfluous (code would work the same without it).

File: source/agora/script/Signature.d [hunk 4]
```diff
@@ -74,23 +98,64 @@ align(1):
 *******************************************************************************/
 
 public string decodeSignature (const(ubyte)[] bytes,
-    out SigPair sig_pair) pure nothrow @safe @nogc
+    out SigPair sig_pair, out ulong pop_count) pure nothrow @safe @nogc
 {
-    if (bytes.length != SigPair.sizeof)
+    pop_count = 0;
```
Comment (Geod24) [inline source/agora/script/Signature.d]: `out` does that for you

File: source/agora/script/Signature.d [hunk 5]
```diff
@@ -74,23 +98,64 @@ align(1):
...
 
-    sig_pair = SigPair(sig, sig_hash);
+    ulong output_idx;
+    if (sig_hash & SigHash.Single && bytes.length >= ulong.sizeof)
+    {
+        output_idx = littleEndianToNative!ulong(bytes[0 .. ulong.sizeof]);
+        pop_count += ulong.sizeof;
+    }
+    else if (sig_hash & SigHash.Single)
+        return "Encoded signature does not have output idx";
```
Comment (Geod24) [inline source/agora/script/Signature.d]:
```suggestion
    if (sig_hash & SigHash.Single)
    {
        if (bytes.length < ulong.sizeof)
            return "Encoded signature does not have output idx";
        output_idx = littleEndianToNative!ulong(bytes[0 .. ulong.sizeof]);
        pop_count += ulong.sizeof;
    }
```

Would be equivalent, right ?

File: source/agora/script/Signature.d [hunk 6]
```diff
@@ -74,23 +98,64 @@ align(1):
...
+    auto serialized = pair[].dup;
+
+    SigPair decoded_pair;
+    ulong consumed;
+
+    assert(decodeSignature(serialized, decoded_pair, consumed) is null);
+    assert(pair == decoded_pair);
+    assert(consumed == serialized.length);
+
+    serialized ~= ubyte(44);
```
Comment (Geod24) [inline source/agora/script/Signature.d]: Don't think you need the `ubyte()` here, the compiler knows the type of the LHS.

File: source/agora/script/Signature.d [hunk 7]
```diff
@@ -142,39 +213,147 @@ unittest
...
-        dup.inputs = dup.inputs.dup;
-        dup.inputs[input_idx] = Input.init;  // blank out matching input
+        dup.inputs = dup.inputs[0..input_idx] ~ dup.inputs[input_idx + 1..$]; // blank out matching input
+        return hashMulti(dup, sig_hash);
+    // sign all inputs and a single output
+    case SigHash.Single:
+        dup.outputs = dup.outputs[output_idx .. output_idx + 1];
+        return hashMulti(dup, sig_hash);
+    // sign a single input and a single output
+    case SigHash.Single | SigHash.AnyoneCanPay:
```
Comment (Geod24) [inline source/agora/script/Signature.d]: One reason why using OOB values for `enum` is bad: The compiler could decide that this is definitely unreachable. Not to mention, `final switch` is useless here.

File: source/agora/test/Flash.d [hunk 8]
```diff
@@ -418,8 +418,9 @@ public class FlashNodeFactory (FlashListenerType = FlashListener)
     public RemoteAPI!TestFlashListenerAPI createFlashListener (
         Listener : TestFlashListenerAPI)(string address)
     {
+        const string agora_address = format("Validator #%s (%s)", 0, genesis_validator_keys[0].address);
         RemoteAPI!TestFlashListenerAPI api
-            = RemoteAPI!TestFlashListenerAPI.spawn!Listener(5.seconds);
+            = RemoteAPI!TestFlashListenerAPI.spawn!Listener(this.agora_registry, agora_address, 5.seconds);
```
Comment (Geod24) [inline source/agora/test/Flash.d:423]: Any hope we can merge this custom impl into `Base.d` ? :D

File: source/agora/flash/Node.d [hunk 9]
```diff
@@ -1292,6 +1295,19 @@ public abstract class FlashNode : FlashControlAPI
...
+    {
+        auto per_byte = this.listener.getEstimatedTxFee();
+        if(!per_byte.mul(tx_size))
+            return FeeUTXOs.init;
+        auto utxos = this.listener.getFeeUTXOs(pk, per_byte);
+        // reuse total_value as refund amount
+        if (!utxos.total_value.sub(per_byte))
+            utxos.total_value = Amount(0);
+        return utxos;
+    }
```
Comment (Geod24) [inline source/agora/flash/Node.d:1310]: Why do we have two implementations of this ?

File: source/agora/flash/UpdateSigner.d (hunk 2 above)
Comment (omerfirmak) [reply source/agora/flash/UpdateSigner.d]: I explicitly defined them.

File: source/agora/flash/Node.d (hunk 9 above)
Comment (omerfirmak) [reply source/agora/flash/Node.d:1310]: this one takes `tx_size`
`FlashListener.getFeeUTXOs` takes flat `Amount`

File: source/agora/script/Signature.d (hunk 1 above)
Comment (omerfirmak) [reply source/agora/script/Signature.d:239]: Deep-dup thru ser/deser is not possible atm.
`TxBuilder` is signing invalid TXs while trying to calculate minimum fees and that triggers assertions in ctors.

https://github.com/bosagora/agora/blob/be48645a0c1e921bb562c5ed10873bf7ceccf3de/source/agora/utils/TxBuilder.d#L598

File: source/agora/test/Flash.d (hunk 8 above)
Comment (omerfirmak) [reply source/agora/test/Flash.d:423]: `FlashNodeFactory` as a whole?

**Pull request merged**